### PR TITLE
NOTIF-324 Allow to filter events by composite types

### PR DESCRIPTION
**Pull request by josejulio**
There is a PR to add some filters on the frontend - one of these is the Splunk integration, which is of type camel and subtype splunk. 

Before this PR, EventLog API did not support searching for sub_types. This PR enables searching for type with optional subtypes.

### CODE REVIEW

File: database/src/main/resources/db/migration/V1.49.0__NOTIF-324_ensure-subtypes-are-lower-case.sql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+-- ensure the sub_types are lower case
+
+UPDATE endpoints
+    SET   endpoint_sub_type = LOWER(endpoint_sub_type)
+    WHERE endpoint_sub_type != LOWER(endpoint_sub_type);
+
+UPDATE notification_history
+    SET   endpoint_sub_type = LOWER(endpoint_sub_type)
+    WHERE endpoint_sub_type != LOWER(endpoint_sub_type)
```
Comment (gwenneg) [inline database/src/main/resources/db/migration/V1.49.0__NOTIF-324_ensure-subtypes-are-lower-case.sql:9]: FYI, I just checked this with gabi:
| | Stage | Prod |
| - | - | - |
| SELECT COUNT(*) FROM endpoints WHERE endpoint_sub_type != LOWER(endpoint_sub_type) | 0 | 0 |
| SELECT COUNT(*) FROM notification_history WHERE endpoint_sub_type != LOWER(endpoint_sub_type) | 0 | 0 |

Comment (josejulio) [reply database/src/main/resources/db/migration/V1.49.0__NOTIF-324_ensure-subtypes-are-lower-case.sql:9]: Ah - then I can remove it:)

Comment (theute) [reply database/src/main/resources/db/migration/V1.49.0__NOTIF-324_ensure-subtypes-are-lower-case.sql:9]: @gwenneg enjoy while it lasts your Gabi days are counted ;-)

> expirationDate: '2022-05-01'

(we can renew)

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java [hunk 2]
```diff
@@ -125,7 +127,19 @@ private static String addHqlConditions(String hql, Set<UUID> bundleIds, Set<UUID
         if (checkEndpointType || checkInvocationResult) {
             List<String> subQueryConditions = new ArrayList<>();
             if (checkEndpointType) {
-                subQueryConditions.add("nh.endpointType IN (:endpointTypes)");
+                List<String> subQueryEndpointTypes = new ArrayList<>();
+                Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
+                Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:132]: With the current code, this is executed 4 times when `EventResource#getEvents` is called. Could you move it to `EventResource` and pass the basic and composite types as method params to `EventRepository`?

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java [hunk 3]
```diff
@@ -125,7 +127,19 @@ private static String addHqlConditions(String hql, Set<UUID> bundleIds, Set<UUID
...
+
+                if (basicTypes.size() > 0) {
+                    subQueryEndpointTypes.add("nh.compositeEndpointType.type IN (:basicEndpointTypes)");
+                }
+
+                if (compositeTypes.size() > 0) {
+                    subQueryEndpointTypes.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
+                }
```
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: Did you check if this subquery works for basic types? `CompositeEndpointType#equals` compares both `type` and `subType`. If `type == null` then this query could still work. I'm not sure of Hibernate behavior in that case, this needs to be confirmed.

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java (hunk 2 above)
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:132]: My [other comment](https://github.com/RedHatInsights/notifications-backend/pull/1183#discussion_r855374697) may change things a bit here.

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java (hunk 3 above)
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: > If `type == null` then this query could still work

I suppose you mean `sub_type`? `type` should never be `null`

I'll double check - I recall I checked this before and when the `sub_type` was `null` it wasn't working well. But I guess the query evolved a bit and is worth to check if that's still needed.

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java (hunk 2 above)
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:132]: Sure

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java (hunk 3 above)
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: Yes, I tried to use the composite directly (with nulls) and it didn't work.
Fails this test:  https://github.com/RedHatInsights/notifications-backend/blob/93159fc5a67bc6e06ab7523822fa08c5075b12ad/backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java#L282

```
Expected :1
Actual   :0
```

```
diff --git a/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java b/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
index 96be5149..05042701 100644
--- a/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
+++ b/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
@@ -127,19 +127,7 @@ public class EventRepository {
         if (checkEndpointType || checkInvocationResult) {
             List<String> subQueryConditions = new ArrayList<>();
             if (checkEndpointType) {
-                List<String> subQueryEndpointTypes = new ArrayList<>();
-                Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
-                Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
-
-                if (basicTypes.size() > 0) {
-                    subQueryEndpointTypes.add("nh.compositeEndpointType.type IN (:basicEndpointTypes)");
-                }
-
-                if (compositeTypes.size() > 0) {
-                    subQueryEndpointTypes.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
-                }
-
-                subQueryConditions.add("(" + String.join(" OR ", subQueryEndpointTypes) + ")");
+                subQueryConditions.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
             }
             if (checkInvocationResult) {
                 subQueryConditions.add("nh.invocationResult IN (:invocationResults)");
@@ -169,16 +157,7 @@ public class EventRepository {
             query.setParameter("endDate", Timestamp.valueOf(endDate.atTime(LocalTime.MAX))); // at end of day
         }
         if (endpointTypes != null && !endpointTypes.isEmpty()) {
-            Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
-            Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
-
-            if (basicTypes.size() > 0) {
-                query.setParameter("basicEndpointTypes", basicTypes);
-            }
-
-            if (compositeTypes.size() > 0) {
-                query.setParameter("compositeEndpointTypes", compositeTypes);
-            }
+            query.setParameter("compositeEndpointTypes", endpointTypes);
         }
         if (invocationResults != null && !invocationResults.isEmpty()) {
             query.setParameter("invocationResults", invocationResults);

```

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: Do you have the SQL generated by Hibernate when that test fails?

Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: Get Ids
```
select
        event0_.id as col_0_0_ 
    from
        event event0_ 
    where
        event0_.account_id=? 
        and (
            event0_.bundle_id in (
                ?
            )
        ) 
        and (
            event0_.application_id in (
                ?
            )
        ) 
        and (
            lower(event0_.event_type_display_name) like ?
        ) 
        and (
            event0_.created between ? and ?
        ) 
        and (
            exists (
                select
                    1 
                from
                    notification_history notificati1_ 
                where
                    notificati1_.event_id=event0_.id 
                    and (
                        (
                            notificati1_.endpoint_sub_type, notificati1_.endpoint_type
                        ) in (
                            (
                                ?, ?
                            )
                        )
                    ) 
                    and (
                        notificati1_.invocation_result in (
                            ?
                        )
                    )
            )
        ) 
    order by
        event0_.created desc limit ?
```

Get events
```
    select
        distinct event0_.id as id1_10_0_,
        historyent1_.id as id1_15_1_,
        event0_.account_id as account_2_10_0_,
        event0_.application_display_name as applicat3_10_0_,
        event0_.application_id as applicat4_10_0_,
        event0_.bundle_display_name as bundle_d5_10_0_,
        event0_.bundle_id as bundle_i6_10_0_,
        event0_.created as created7_10_0_,
        event0_.event_type_id as event_t10_10_0_,
        event0_.event_type_display_name as event_ty8_10_0_,
        event0_.payload as payload9_10_0_,
        historyent1_.created as created2_15_1_,
        historyent1_.endpoint_sub_type as endpoint3_15_1_,
        historyent1_.endpoint_type as endpoint4_15_1_,
        historyent1_.details as details5_15_1_,
        historyent1_.endpoint_id as endpoint8_15_1_,
        historyent1_.event_id as event_id9_15_1_,
        historyent1_.invocation_result as invocati6_15_1_,
        historyent1_.invocation_time as invocati7_15_1_,
        historyent1_.event_id as event_id9_15_0__,
        historyent1_.id as id1_15_0__ 
    from
        event event0_ 
    left outer join
        notification_history historyent1_ 
            on event0_.id=historyent1_.event_id 
    where
        event0_.id in (
            null
        ) 
    order by
        event0_.created desc

```

Count
```    
select
        count(*) as col_0_0_ 
    from
        event event0_ 
    where
        event0_.account_id=? 
        and (
            event0_.bundle_id in (
                ?
            )
        ) 
        and (
            event0_.application_id in (
                ?
            )
        ) 
        and (
            lower(event0_.event_type_display_name) like ?
        ) 
        and (
            event0_.created between ? and ?
        ) 
        and (
            exists (
                select
                    1 
                from
                    notification_history notificati1_ 
                where
                    notificati1_.event_id=event0_.id 
                    and (
                        (
                            notificati1_.endpoint_sub_type, notificati1_.endpoint_type
                        ) in (
                            (
                                ?, ?
                            )
                        )
                    ) 
                    and (
                        notificati1_.invocation_result in (
                            ?
                        )
                    )
            )
        )
```

Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java:140]: I see why it doesn't work: `something IN (null)` doesn't work with Postgres, so the query generated by Hibernate will never work. Writing a native query for that would be overkill, let's keep the current code.

File: backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java (hunk 2 above)
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java]: done

File: backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java [hunk 4]
```diff
@@ -374,6 +377,53 @@ void testAllQueryParams() {
...
+         */
+        page = getEventLogPage(defaultIdentityHeader, null, null, null, null, null, Set.of("camel:slack"), null, null, null, null, false, true);
+        assertEquals(1, page.getMeta().getCount());
+        assertEquals(1, page.getData().size());
+        assertSameEvent(page.getData().get(0), event3, history4, history5);
+        assertNull(page.getData().get(0).getPayload());
+        assertLinks(page.getLinks(), "first", "last");
+
+        /*
+         * Test #24
```
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java:416]:
```suggestion
         * Test #25
```

**Review by gwenneg — COMMENTED** [review]
LGTM

File: backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java (hunk 4 above)
Comment (gwenneg) [inline backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java:416]:
```suggestion
         * Test #25
```

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**